agent/engine/task_manager.go

```diff
@@ -732,8 +823,83 @@ func (mtask *managedTask) startContainerTransitions(transitionFunc containerTran
 	return anyCanTransition, transitions, reasons
 }
 
+// startResourceTransitions steps through each resource in the task and calls
+// the passed transition function when a transition should occur
+func (mtask *managedTask) startResourceTransitions(transitionFunc resourceTransitionFunc) (bool, map[string]string) {
+	anyCanTransition := false
+	transitions := make(map[string]string)
+	for _, res := range mtask.Resources {
+		knownStatus := res.GetKnownStatus()
+		desiredStatus := res.GetDesiredStatus()
+		if knownStatus >= desiredStatus {
+			seelog.Debugf("Managed task [%s]: resource [%s] has already transitioned to or beyond the desired status %s; current known is %s",
+				mtask.Arn, res.GetName(), res.StatusString(desiredStatus), res.StatusString(knownStatus))
+			continue
+		}
+		anyCanTransition = true
+		transition := mtask.resourceNextState(res)
+		// If the resource is already in a transition, skip
+		if transition.actionRequired && !res.SetAppliedStatus(transition.nextState) {
+			// At least one resource is able to be moved forwards, so we're not deadlocked
+			continue
+		}
+		if !transition.actionRequired {
+			// no action is required for the transition, just set the known status without
+			// calling any transition function
+			go func(res taskresource.TaskResource, status taskresource.ResourceStatus, err error) {
+				mtask.resourceStateChangeEvent <- resourceStateChange{
+					resource:  res,
+					nextState: transition.nextState,
+					err:       nil,
+				}
+			}(res, transition.nextState, nil)
+			continue
+		}
+		// At least one resource is able to be move forwards, so we're not deadlocked
+		transitions[res.GetName()] = transition.status
+		go transitionFunc(res, transition.nextState)
+	}
+
+	return anyCanTransition, transitions
+}
+
+// transitionResource calls applyResourceState, and then notifies the managed
+// task of the change. transitionResource is called by progressTask
+func (mtask *managedTask) transitionResource(resource taskresource.TaskResource,
+	to taskresource.ResourceStatus) {
+	engine := mtask.engine
+	err := mtask.applyResourceState(resource, to)
+
+	if engine.IsTaskManaged(mtask.Arn) {
+		mtask.handleResourceStateChange(resourceStateChange{
```

Comment:
This still shouldn't be called directly, an event should be sent to `resourceStateChangeEvent` instead. And instead of check `IsTaskManaged` you can do something like [this](https://github.com/aws/amazon-ecs-agent/pull/1323/files#diff-20346f7af6202828de353f0137b6e43aR454)